plugins/nf-validation/src/main/nextflow/validation/SamplesheetConverter.groovy

```diff
@@ -0,0 +1,327 @@
+                    }
+                }
+                
+                // Check if the field is unique
+                def unique = field['value']['unique']
+                def Boolean uniqueIsList = unique instanceof ArrayList
+                if(unique && !uniqueIsList){
+                    if(!(key in booleanUniques)){
+                        booleanUniques[key] = []
+                    }
+                    if(input in booleanUniques[key] && input){
+                        this.errors << "The '${key}' value needs to be unique. '${input}' was found at least twice in the samplesheet.".toString()
+                    }
+                    booleanUniques[key].add(input)
+                }
+                else if(unique && uniqueIsList) {
+                    def Map<String,String> newMap = row.subMap([key] + (List) unique)
+                    if(!(key in listUniques)){
+                        listUniques[key] = []
+                    }
+                    if(newMap in listUniques[key] && input){
+                        this.errors << "The combination of '${key}' with fields ${unique} needs to be unique. ${newMap} was found at least twice.".toString()
+                    }
+                    listUniques[key].add(newMap)
+                }
+
+                // Check enumeration
+                def List enumeration = field['value']['enum'] as List
+                if(input && enumeration && !(enumeration.contains(input))){
+                    this.errors << "The '${key}' value for sample ${this.getCount()} needs to be one of ${enumeration}, but is '${input}'.".toString()
+                }
+
+                // Convert field to a meta field or add it as an input to the channel
+                def String metaNames = field['value']['meta']
+                if(metaNames) {
+                    for(name : metaNames.tokenize(',')) {
+                        meta[name] = (input != '' && input) ? 
+                                checkAndTransform(input, field) : 
+                            field['value']['default'] ? 
+                                checkAndTransform(field['value']['default'] as String, field) : 
+                                null
+                    }
+                }
+                else {
+                    def inputFile = (input != '' && input) ? 
+                            checkAndTransform(input, field) : 
+                        field['value']['default'] ? 
+                            checkAndTransform(field['value']['default'] as String, field) : 
+                            []
+                    output.add(inputFile)
+                }
+            }
+            output.add(0, meta)
+            return output
+        }
+
+        // check for errors
+        if (this.hasErrors()) {
+            String message = this.getErrors().join("\n").trim() + this.getSchemaErrors().join("\n").trim()
```

Comment:
Could we group the errors? 
Thinking on something like:
```
Samplesheet Errors: 
    The 'cram' file or directory (assets/test.cram) for sample 1 does not exist.
```
Instead of repeating the "[Samplesheet Errors]" tag for every message. Similar to what we do with parameter validation.

Reply:
Also, I did a test and got
```
[Samplesheet Error] The 'cram' value (assets/test.cram) for sample 1 is not a valid integer.
[Samplesheet Error] The 'crai' file or directory (assets/test.cram.crai) for sample 1 does not exist.
[Samplesheet Error] The 'ped' file or directory (assets/test.ped) for sample 1 does not exist.
[Samplesheet Error] The 'cram' value (assets/test.cram) for sample 2 is not a valid integer.
[Samplesheet Error] The 'bed' file or directory (assets/test.bed) for sample 2 does not exist.
[Samplesheet Error] The 'ped' file or directory (assets/test.ped) for sample 2 does not exist.
```
Maybe we could also sort the messages. By type of error (invalid value, file not found...) or by column (cram, crai...)

Reply:
Oh yes that's a way better error handling, I'll fix that, thanks for the feedback!

Reply:
I concatenated the errors like you showed in the first comment here. Furthermore I added a `.sort()` to the errors (it's not perfect but should already make it a bit better, I'll improve this more after moving the validation code to the library)